File: packages/python/src/brainpick/detect.py
"""Environment detection for init/doctor (docs/embedding-detection.md, docs/onboarding.md).

Detect rather than interrogate: bundle shape and link style come from reading the
markdown, backends from parallel 300 ms probes. A probe that fails is a silent miss —
detection never raises and never stalls the choreography.
"""
from __future__ import annotations

import os
import shutil
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from pathlib import Path
from typing import Mapping

import httpx

from brainpick.core.bundle import ALWAYS_EXCLUDED_DIRS
from brainpick.core.frontmatter import split_frontmatter
from brainpick.core.links import extract_links

PROBE_TIMEOUT = 0.3  # seconds; a miss must never make init feel slow
PREFERRED_EMBEDDING_MODELS = (
    "nomic-embed-text",
    "mxbai-embed-large",
    "snowflake-arctic-embed2",
    "bge-m3",
)
DEFAULT_OLLAMA = "http://127.0.0.1:11434"
DEFAULT_OPENAI_COMPATIBLE = (
    ("lm studio", "http://127.0.0.1:1234"),
    ("llama.cpp", "http://127.0.0.1:8080"),
)
MIN_TYPED_DOCS = 3  # density scan: this many `type:` frontmatters look like a bundle


@dataclass(frozen=True)
class Backend:
    kind: str  # "ollama" | "openai-compatible" | "openai"
    endpoint: str
    model: str | None  # None: the endpoint answered but offers no embedding model


@dataclass(frozen=True)
class BundleInfo:
    kind: str  # "okf" | "density" | "none"
    docs: int  # markdown files seen (excluded dirs skipped)
    typed: int  # of those, files with `type:` frontmatter


@dataclass(frozen=True)
class LinkStyle:
    style: str  # "markdown" | "wikilinks" | "mixed" | "none"
    markdown: int
    wikilinks: int


# -- bundle ------------------------------------------------------------------------


def _markdown_files(root: Path) -> list[Path]:
    files = []
    for path in sorted(root.rglob("*.md")):
        parts = path.relative_to(root).parts
        if any(part in ALWAYS_EXCLUDED_DIRS for part in parts[:-1]):
            continue
        if path.is_file():
            files.append(path)
    return files


def detect_bundle(root: str | Path) -> BundleInfo:
    """okf (index.md declares okf_version) > density (>= 3 typed docs) > none."""
    root = Path(root)
    files = _markdown_files(root)
    okf = False
    typed = 0
    for path in files:
        meta, _body = split_frontmatter(path.read_text(encoding="utf-8", errors="replace"))
        if path == root / "index.md" and meta.get("okf_version") is not None:
            okf = True
        if meta.get("type") is not None:
            typed += 1
    if okf:
        return BundleInfo("okf", len(files), typed)
    if typed >= MIN_TYPED_DOCS:
        return BundleInfo("density", len(files), typed)
    return BundleInfo("none", len(files), typed)


def detect_link_style(root: str | Path) -> LinkStyle:
    """Informational in 0.1: how this bundle links, counted from the bodies."""
    root = Path(root)
    markdown = wikilinks = 0
    for path in _markdown_files(root):
        _meta, body = split_frontmatter(path.read_text(encoding="utf-8", errors="replace"))
        for link in extract_links(body):
            if link.kind == "wikilink":
                wikilinks += 1
            else:
                markdown += 1
    if markdown == 0 and wikilinks == 0:
        style = "none"
    elif wikilinks == 0:
        style = "markdown"
    elif markdown == 0:
        style = "wikilinks"
    else:
        style = "mixed"
    return LinkStyle(style, markdown, wikilinks)


# -- backends ----------------------------------------------------------------------


def _get_json(url: str) -> dict | None:
    try:
        response = httpx.get(url, timeout=PROBE_TIMEOUT)
        response.raise_for_status()
        data = response.json()
    except Exception:
        return None  # a miss is silent — down, slow, or gibberish all mean "not here"
    return data if isinstance(data, dict) else None


def _pick_embedding_model(names: list[str]) -> str | None:
    embeddable = [name for name in names if "embed" in name.lower()]
    for preferred in PREFERRED_EMBEDDING_MODELS:
        for name in embeddable:
            if preferred in name:
                return name
    return embeddable[0] if embeddable else None


def _normalize_host(value: str) -> str:
    value = value.strip().rstrip("/")
    if "://" not in value:
        value = f"http://{value}"
    return value


def probe_ollama(env: Mapping[str, str] | None = None) -> Backend | None:
    env = os.environ if env is None else env
    base = _normalize_host(env.get("OLLAMA_HOST") or DEFAULT_OLLAMA)
    data = _get_json(f"{base}/api/tags")
    if data is None:
        return None
    names = [str(m.get("name", "")) for m in data.get("models", []) if isinstance(m, dict)]
    return Backend("ollama", base, _pick_embedding_model(names))


def probe_openai_compatible(base: str) -> Backend | None:
    base = _normalize_host(base)
    data = _get_json(f"{base}/v1/models")
    if data is None:
        return None
    ids = [str(m.get("id", "")) for m in data.get("data", []) if isinstance(m, dict)]
    return Backend("openai-compatible", f"{base}/v1", _pick_embedding_model(ids))


def probe_backends(
    env: Mapping[str, str] | None = None,
    openai_compatible: tuple[tuple[str, str], ...] = DEFAULT_OPENAI_COMPATIBLE,
) -> list[tuple[str, Backend | None]]:
    """All probes in parallel, ladder order preserved: ollama, then OpenAI-compatible."""
    env = os.environ if env is None else env
    with ThreadPoolExecutor(max_workers=1 + len(openai_compatible)) as pool:
        first = pool.submit(probe_ollama, env)
        rest = [(label, pool.submit(probe_openai_compatible, base)) for label, base in openai_compatible]
        results: list[tuple[str, Backend | None]] = [("ollama", first.result())]
        results.extend((label, future.result()) for label, future in rest)
    return results


def pick_backend(results: list[tuple[str, Backend | None]]) -> Backend | None:
    """The first backend that actually has an embedding model — the ladder's answer."""
    for _label, backend in results:
        if backend is not None and backend.model is not None:
            return backend
    return None


def openai_key_present(env: Mapping[str, str] | None = None) -> bool:
    env = os.environ if env is None else env
    return bool(env.get("OPENAI_API_KEY"))


# -- surroundings ------------------------------------------------------------------


def find_repo_root(start: str | Path) -> Path | None:
    """The nearest ancestor (or self) holding a .git — where .gitignore would live."""
    path = Path(start).resolve()
    for candidate in (path, *path.parents):
        if (candidate / ".git").exists():
            return candidate
    return None


def detect_henxels(root: str | Path) -> Path | None:
    """henxels.yaml at the bundle root, or at the repo root above it."""
    root = Path(root).resolve()
    contract = root / "henxels.yaml"
    if contract.is_file():
        return contract
    repo = find_repo_root(root)
    if repo is not None and repo != root:
        contract = repo / "henxels.yaml"
        if contract.is_file():
            return contract
    return None


def henxels_on_path() -> bool:
    return shutil.which("henxels") is not None

File: packages/node/src/detect.ts
/** Environment detection for init/doctor (docs/embedding-detection.md, docs/onboarding.md).
 *
 * Detect rather than interrogate: bundle shape and link style come from reading the
 * markdown, backends from parallel 300 ms probes. A probe that fails is a silent miss —
 * detection never raises and never stalls the choreography. Ports detect.py.
 */
import { accessSync, constants, readdirSync, readFileSync, statSync } from "node:fs";
import { delimiter, join, resolve } from "node:path";

import { ALWAYS_EXCLUDED_DIRS } from "./core/bundle";
import { cmpStr } from "./core/canonical";
import { splitFrontmatter } from "./core/frontmatter";
import { extractLinks } from "./core/links";

export const PROBE_TIMEOUT_MS = 300; // a miss must never make init feel slow
export const PREFERRED_EMBEDDING_MODELS = [
  "nomic-embed-text",
  "mxbai-embed-large",
  "snowflake-arctic-embed2",
  "bge-m3",
] as const;
export const DEFAULT_OLLAMA = "http://127.0.0.1:11434";
export const DEFAULT_OPENAI_COMPATIBLE: ReadonlyArray<readonly [string, string]> = [
  ["lm studio", "http://127.0.0.1:1234"],
  ["llama.cpp", "http://127.0.0.1:8080"],
];
export const MIN_TYPED_DOCS = 3; // density scan: this many `type:` frontmatters look like a bundle

export type Env = Record<string, string | undefined>;

export interface Backend {
  kind: string; // "ollama" | "openai-compatible" | "openai"
  endpoint: string;
  model: string | null; // null: the endpoint answered but offers no embedding model
}

export interface BundleInfo {
  kind: string; // "okf" | "density" | "none"
  docs: number; // markdown files seen (excluded dirs skipped)
  typed: number; // of those, files with `type:` frontmatter
}

export interface LinkStyle {
  style: string; // "markdown" | "wikilinks" | "mixed" | "none"
  markdown: number;
  wikilinks: number;
}

export type ProbeResult = readonly [label: string, backend: Backend | null];

// -- bundle ------------------------------------------------------------------------

function markdownFiles(root: string): string[] {
  const out: string[] = [];
  const walk = (dir: string, rel: string): void => {
    let entries;
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      const childRel = rel === "" ? entry.name : `${rel}/${entry.name}`;
      if (entry.isDirectory()) {
        if (!ALWAYS_EXCLUDED_DIRS.has(entry.name)) walk(join(dir, entry.name), childRel);
      } else if (entry.name.endsWith(".md")) {
        let isFile = entry.isFile();
        if (entry.isSymbolicLink()) {
          try {
            isFile = statSync(join(dir, entry.name)).isFile();
          } catch {
            isFile = false;
          }
        }
        if (isFile) out.push(childRel);
      }
    }
  };
  walk(root, "");
  return out.sort(cmpStr);
}

/** okf (index.md declares okf_version) > density (>= 3 typed docs) > none. */
export function detectBundle(root: string): BundleInfo {
  const files = markdownFiles(root);
  let okf = false;
  let typed = 0;
  for (const rel of files) {
    let text: string;
    try {
      text = readFileSync(join(root, rel), "utf8");
    } catch {
      continue;
    }
    const [meta] = splitFrontmatter(text);
    if (rel === "index.md" && meta["okf_version"] !== null && meta["okf_version"] !== undefined) {
      okf = true;
    }
    if (meta["type"] !== null && meta["type"] !== undefined) typed += 1;
  }
  if (okf) return { kind: "okf", docs: files.length, typed };
  if (typed >= MIN_TYPED_DOCS) return { kind: "density", docs: files.length, typed };
  return { kind: "none", docs: files.length, typed };
}

/** Informational in 0.1: how this bundle links, counted from the bodies. */
export function detectLinkStyle(root: string): LinkStyle {
  let markdown = 0;
  let wikilinks = 0;
  for (const rel of markdownFiles(root)) {
    let text: string;
    try {
      text = readFileSync(join(root, rel), "utf8");
    } catch {
      continue;
    }
    const [, body] = splitFrontmatter(text);
    for (const link of extractLinks(body)) {
      if (link.kind === "wikilink") wikilinks += 1;
      else markdown += 1;
    }
  }
  let style: string;
  if (markdown === 0 && wikilinks === 0) style = "none";
  else if (wikilinks === 0) style = "markdown";
  else if (markdown === 0) style = "wikilinks";
  else style = "mixed";
  return { style, markdown, wikilinks };
}

// -- backends ----------------------------------------------------------------------

async function getJson(url: string): Promise<Record<string, unknown> | null> {
  try {
    const response = await fetch(url, { signal: AbortSignal.timeout(PROBE_TIMEOUT_MS) });
    if (!response.ok) return null;
    const data: unknown = await response.json();
    return typeof data === "object" && data !== null && !Array.isArray(data)
      ? (data as Record<string, unknown>)
      : null;
  } catch {
    return null; // a miss is silent — down, slow, or gibberish all mean "not here"
  }
}

function pickEmbeddingModel(names: string[]): string | null {
  const embeddable = names.filter((name) => name.toLowerCase().includes("embed"));
  for (const preferred of PREFERRED_EMBEDDING_MODELS) {
    for (const name of embeddable) {
      if (name.includes(preferred)) return name;
    }
  }
  return embeddable[0] ?? null;
}

function normalizeHost(value: string): string {
  let host = value.trim().replace(/\/+$/, "");
  if (!host.includes("://")) host = `http://${host}`;
  return host;
}

export async function probeOllama(env: Env = process.env): Promise<Backend | null> {
  const base = normalizeHost(env["OLLAMA_HOST"] || DEFAULT_OLLAMA);
  const data = await getJson(`${base}/api/tags`);
  if (data === null) return null;
  const models = Array.isArray(data["models"]) ? (data["models"] as unknown[]) : [];
  const names = models
    .filter((m): m is Record<string, unknown> => typeof m === "object" && m !== null && !Array.isArray(m))
    .map((m) => String(m["name"] ?? ""));
  return { kind: "ollama", endpoint: base, model: pickEmbeddingModel(names) };
}

export async function probeOpenaiCompatible(base: string): Promise<Backend | null> {
  const host = normalizeHost(base);
  const data = await getJson(`${host}/v1/models`);
  if (data === null) return null;
  const models = Array.isArray(data["data"]) ? (data["data"] as unknown[]) : [];
  const ids = models
    .filter((m): m is Record<string, unknown> => typeof m === "object" && m !== null && !Array.isArray(m))
    .map((m) => String(m["id"] ?? ""));
  return { kind: "openai-compatible", endpoint: `${host}/v1`, model: pickEmbeddingModel(ids) };
}

/** All probes in parallel, ladder order preserved: ollama, then OpenAI-compatible. */
export async function probeBackends(
  env: Env = process.env,
  openaiCompatible: ReadonlyArray<readonly [string, string]> = DEFAULT_OPENAI_COMPATIBLE,
): Promise<ProbeResult[]> {
  const first = probeOllama(env);
  const rest = openaiCompatible.map(
    async ([label, base]) => [label, await probeOpenaiCompatible(base)] as const,
  );
  return [["ollama", await first] as const, ...(await Promise.all(rest))];
}

/** The first backend that actually has an embedding model — the ladder's answer. */
export function pickBackend(results: readonly ProbeResult[]): Backend | null {
  for (const [, backend] of results) {
    if (backend !== null && backend.model !== null) return backend;
  }
  return null;
}

export function openaiKeyPresent(env: Env = process.env): boolean {
  return Boolean(env["OPENAI_API_KEY"]);
}

// -- surroundings ------------------------------------------------------------------

/** The nearest ancestor (or self) holding a .git — where .gitignore would live. */
export function findRepoRoot(start: string): string | null {
  let candidate = resolve(start);
  for (;;) {
    try {
      statSync(join(candidate, ".git"));
      return candidate;
    } catch {
      /* keep climbing */
    }
    const parent = resolve(candidate, "..");
    if (parent === candidate) return null;
    candidate = parent;
  }
}

/** henxels.yaml at the bundle root, or at the repo root above it. */
export function detectHenxels(root: string): string | null {
  const resolved = resolve(root);
  const local = join(resolved, "henxels.yaml");
  try {
    if (statSync(local).isFile()) return local;
  } catch {
    /* not here */
  }
  const repo = findRepoRoot(resolved);
  if (repo !== null && repo !== resolved) {
    const contract = join(repo, "henxels.yaml");
    try {
      if (statSync(contract).isFile()) return contract;
    } catch {
      /* not there either */
    }
  }
  return null;
}

/** shutil.which, minimally: scan PATH for an executable regular file. */
export function which(cmd: string, env: Env = process.env): string | null {
  if (cmd.includes("/") || cmd.includes("\\")) {
    return isExecutableFile(cmd) ? cmd : null;
  }
  const paths = (env["PATH"] ?? "").split(delimiter);
  const exts =
    process.platform === "win32" ? (env["PATHEXT"] ?? ".COM;.EXE;.BAT;.CMD").split(";") : [""];
  for (const dir of paths) {
    if (dir === "") continue;
    for (const ext of exts) {
      const candidate = join(dir, cmd + ext.toLowerCase());
      if (isExecutableFile(candidate)) return candidate;
    }
  }
  return null;
}

function isExecutableFile(path: string): boolean {
  try {
    if (!statSync(path).isFile()) return false;
    accessSync(path, constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

export function henxelsOnPath(env: Env = process.env): boolean {
  return which("henxels", env) !== null;
}
